### ISSUE TITLE
cannot remove @timestamp field

### ISSUE THREAD

**Issue by higee** [logstash-plugins/logstash-output-mongodb#49]
Please post all product and debugging questions on our [forum](https://discuss.elastic.co/c/logstash). Your questions will reach our wider community members there, and if we confirm that there is a bug, then we can open a new issue here.

For all general issues, please provide the following details for fast resolution:

- Version: `logstash 6.1.1`
- Operating System: `Amazon Linux AMI`
- Config File (if you have sensitive info, please remove it): `test.conf`
```
input {
  stdin {
  }
}

filter {
  mutate {
    remove_field => ["@timestamp"]
  }
}

output {
    mongodb {
      uri => ...
      database => ...
      collection => ...
    }
}
```
- Sample Data: just type whatever message to stdin after running logstash
- Steps to Reproduce: 
```
$ bin/logstash -f test.conf
> hello
>>>
[2018-07-26T05:47:05,981][WARN ][logstash.outputs.mongodb ] Failed to send event to MongoDB 
{:event=>#<LogStash::Event:0x6c6dbc2b>, :exception=>#<NoMethodError: undefined method `to_json' for nil:NilClass
Did you mean?  to_bson>, :backtrace=>[
  "/home/ec2-user/elastic/logstash-6.1.1/vendor/bundle/jruby/2.3.0/gems/logstash-output-mongodb-3.1.4/lib/logstash/outputs/mongodb.rb:81:in `receive'", 
  "/home/ec2-user/elastic/logstash-6.1.1/logstash-core/lib/logstash/outputs/base.rb:92:in `block in multi_receive'", "org/jruby/RubyArray.java:1734:in `each'", 
  "/home/ec2-user/elastic/logstash-6.1.1/logstash-core/lib/logstash/outputs/base.rb:92:in `multi_receive'", 
  "/home/ec2-user/elastic/logstash-6.1.1/logstash-core/lib/logstash/output_delegator_strategies/legacy.rb:22:in `multi_receive'", 
  "/home/ec2-user/elastic/logstash-6.1.1/logstash-core/lib/logstash/output_delegator.rb:50:in `multi_receive'", 
  "/home/ec2-user/elastic/logstash-6.1.1/logstash-core/lib/logstash/pipeline.rb:487:in `block in output_batch'", "org/jruby/RubyHash.java:1343:in `each'", 
  "/home/ec2-user/elastic/logstash-6.1.1/logstash-core/lib/logstash/pipeline.rb:486:in `output_batch'", 
  "/home/ec2-user/elastic/logstash-6.1.1/logstash-core/lib/logstash/pipeline.rb:438:in `worker_loop'", 
  "/home/ec2-user/elastic/logstash-6.1.1/logstash-core/lib/logstash/pipeline.rb:393:in `block in start_workers'"
]}
```

**Comment by HansG89** [comment]
Same here but with different error message:

`[2018-10-24T11:05:48,590][WARN ][logstash.outputs.mongodb ] Failed to send event to MongoDB {:event=>#<LogStash::Event:0x2bf44f1d>, :exception=>java.lang.NullPointerException, :backtrace=>["org.jruby.ir.targets.InvokeSite.pollAndGetClass(InvokeSite.java:444)", "org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:125)", "usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.logstash_minus_output_minus_mongodb_minus_3_dot_1_dot_4.lib.logstash.outputs.mongodb.RUBY$method$receive$0(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-output-mongodb-3.1.4/lib/logstash/outputs/mongodb.rb:81)", "usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.logstash_minus_output_minus_mongodb_minus_3_dot_1_dot_4.lib.logstash.outputs.mongodb.RUBY$method$receive$0$__VARARGS__(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-output-mongodb-3.1.4/lib/logstash/outputs/mongodb.rb)", "org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:77)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:93)", "org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:145)", "usr.share.logstash.logstash_minus_core.lib.logstash.outputs.base.RUBY$block$multi_receive$1(/usr/share/logstash/logstash-core/lib/logstash/outputs/base.rb:89)", "org.jruby.runtime.CompiledIRBlockBody.yieldDirect(CompiledIRBlockBody.java:156)", "org.jruby.runtime.BlockBody.yield(BlockBody.java:114)", "org.jruby.runtime.Block.yield(Block.java:165)", "org.jruby.RubyArray.each(RubyArray.java:1734)", "org.jruby.RubyArray$INVOKER$i$0$0$each.call(RubyArray$INVOKER$i$0$0$each.gen)", "org.jruby.internal.runtime.methods.JavaMethod$JavaMethodZeroBlock.call(JavaMethod.java:498)", "org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:145)", "usr.share.logstash.logstash_minus_core.lib.logstash.outputs.base.RUBY$method$multi_receive$0(/usr/share/logstash/logstash-core/lib/logstash/outputs/base.rb:89)", "org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:103)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:163)", "org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:200)", "org.logstash.config.ir.compiler.OutputStrategyExt$AbstractOutputStrategyExt.invokeOutput(OutputStrategyExt.java:124)", "org.logstash.config.ir.compiler.OutputStrategyExt$LegacyOutputStrategyExt.output(OutputStrategyExt.java:188)", "org.logstash.config.ir.compiler.OutputStrategyExt$AbstractOutputStrategyExt.multiReceive(OutputStrategyExt.java:114)", "org.logstash.config.ir.compiler.OutputDelegatorExt.doOutput(OutputDelegatorExt.java:78)", "org.logstash.config.ir.compiler.AbstractOutputDelegatorExt.multiReceive(AbstractOutputDelegatorExt.java:97)", "org.logstash.config.ir.compiler.AbstractOutputDelegatorExt$INVOKER$i$1$0$multiReceive.call(AbstractOutputDelegatorExt$INVOKER$i$1$0$multiReceive.gen)", "org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:739)", "org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:145)", "usr.share.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$block$output_batch$3(/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:373)", "org.jruby.runtime.CompiledIRBlockBody.yieldDirect(CompiledIRBlockBody.java:156)", "org.jruby.runtime.BlockBody.yield(BlockBody.java:114)", "org.jruby.runtime.Block.yield(Block.java:165)", "org.jruby.RubyHash$12.visit(RubyHash.java:1362)", "org.jruby.RubyHash$12.visit(RubyHash.java:1359)", "org.jruby.RubyHash.visitLimited(RubyHash.java:662)", "org.jruby.RubyHash.visitAll(RubyHash.java:647)", "org.jruby.RubyHash.iteratorVisitAll(RubyHash.java:1319)", "org.jruby.RubyHash.each_pairCommon(RubyHash.java:1354)", "org.jruby.RubyHash.each(RubyHash.java:1343)", "org.jruby.RubyHash$INVOKER$i$0$0$each.call(RubyHash$INVOKER$i$0$0$each.gen)", "org.jruby.internal.runtime.methods.JavaMethod$JavaMethodZeroBlock.call(JavaMethod.java:498)", "org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:145)", "usr.share.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$method$output_batch$0(/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:372)", "usr.share.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$method$output_batch$0$__VARARGS__(/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb)", "org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:77)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:93)", "org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:145)", "usr.share.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$method$worker_loop$0(/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:324)", "usr.share.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$method$worker_loop$0$__VARARGS__(/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb)", "org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:77)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:93)", "org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:145)", "usr.share.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$block$start_workers$2(/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:286)", "org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:145)", "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:71)", "org.jruby.runtime.Block.call(Block.java:124)", "org.jruby.RubyProc.call(RubyProc.java:289)", "org.jruby.RubyProc.call(RubyProc.java:246)", "org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:104)", "java.lang.Thread.run(Thread.java:748)"]}`

**Comment by colinsurprenant** [comment]
Thanks for the report @higee and @HansG89.

Yes, the way the mongodb output was written the `timestamp` field is necessary, per https://github.com/logstash-plugins/logstash-output-mongodb/blob/1c8b4b3de84066c5e1755d0aadf8a535f8196fb6/lib/logstash/outputs/mongodb.rb#L81

Not sure how we could achieve what you want without modifying the plugin. Let me think about it.

**Comment by iurisilvio** [comment]
Setting `isodate => true` fixed my issue. but I think it happens just because it doesn't hit this code line, maybe it has unexpected side effects.

**Comment by colinsurprenant** [comment]
@iurisilvio yeah, setting `isodate => true` skips explicitly setting the mongo document`@timestamp` field from the event `@timestamp` so if/when the event `@timestamp` is missing, the mongo document will not have a `@timestamp` set either.

**Comment by colinsurprenant** [comment]
Removing the event `@timestamp` field using the mutate filter is not a good idea because it is a field which is expected to always be present and some plugins rely on this. That being said, this plugin here could benefit from a better error message for this situation.

**Comment by colinsurprenant** [comment]
improved handling of timestamp and other fixes in #57

**Comment by colinsurprenant** [comment]
should be solved by #57

**Comment by nedngo** [comment]
> Setting `isodate => true` fixed my issue. but I think it happens just because it doesn't hit this code line, maybe it has unexpected side effects.

fixed it
`output {
  mongodb {
    ...
    isodate => true
  }
}
`
